### PR TITLE
Aha integration

### PR DESCRIPTION
**Pull request by thefrieddan1**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
Create Aha! integration
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-2149

## Description
Add integration to enable editing ( closing ) of features in Aha! service. 

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Review by ilappe — CHANGES_REQUESTED** [review]
**very nice job!**
please see my comments

in addition:

- Packs/AHA/README.md - please fill in the file see examples in other packs what should be writen here
- remove the file Packs/AHA/Integrations/AHA/test_data/baseintegration-dummy.json
- Should we have a test playbook for this integration ?
- Please remove all the sensitive data from the test files, use generic data instead.
- Packs/AHA/Author_image.png - please use the same icon as you used for the integration

i still not check the *.py files
will do latter

File: Packs/AHA/Integrations/AHA/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+#### Base Command
+
+`aha-get-features`
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| from_date | Show features created after that date. Default is 2020-01-01. | Optional | 
+| feature_name | get a specific feature. Possible values are: . | Optional |
```
Comment (ilappe) [inline Packs/AHA/Integrations/AHA/README.md:34]: `possible values are` ? :)

File: Packs/AHA/Integrations/AHA/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| feature.id | UUID | The feature ID. | 
+| feature.name | String | The feature name. | 
+| feature.reference_num | String | A feature reference num unique. | 
+| feature.workflow_status | String | Feature status description. |
```
Comment (ilappe) [inline Packs/AHA/Integrations/AHA/README.md:47]: the path aren't updated 
please update the readme according your last changes

**Review by ilappe — CHANGES_REQUESTED** [review]
Great job Dany !!

File: Packs/AHA/pack_metadata.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    "url": "https://www.paloaltonetworks.com/cortex",
+    "email": "",
+    "categories": [
+        "Utilities"
+    ],
+    "tags": [],
+    "useCases": [],
+    "keywords": [],
+    "marketplaces": [
+        "xsoar"
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/pack_metadata.json]: added marketplacev2

File: Packs/AHA/Integrations/AHA/AHA.yml [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+          description: list features starting from a date
+          required: false
+          isArray: false
+          defaultValue: 2020-01-01
+          predefined:
+            - ""
+      outputs: [{
+                "features": [
+                    {
+                        "id": "7142047390424612198",
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.yml]: done

File: Packs/AHA/Integrations/AHA/AHA.yml [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    defaultvalue: ''
+    type: 8
+    additionalinfo: null
+  - name: api_key
+    display: Api Key
+    required: true
+    defaultvalue: null
+    type: 0
+    additionalinfo: "API Key to access service REST API "
+description: Aha!
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.yml]: done

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+
+class Client(BaseClient):
+    def __init__(self, 
+                 headers: dict, 
+                 base_url: str, 
+                 proxy: bool, 
+                 verify: bool):
+        super().__init__(base_url=base_url, proxy=proxy, verify=verify, headers=headers)
+        self._headers = headers
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:27]: right.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+            fields: str optional feature fields to retrive from service
+            from_date: str format: YYYY-MM-DD get feature created after from_date
+            page: str pagination specify the number of the page
+            per_page: str pagination specify the number of maximum features per page.
+        """
+        headers = self._headers
+        url_suffix = f'{URL_SUFFIX}{feature_name}?updated_since={from_date}&fields={fields}&page={page}&per_page={per_page}'
+        return self._http_request(method='GET',
+                                  url_suffix=url_suffix,
+                                  headers=headers, resp_type='json')
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:48]: done.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        desc = fields.get('description')
+        status = fields.get('status')
+        payload = {'feature': {'name': name, 'descriptio': desc,
+                   'workflow_status': {'name': status}}}
+        demisto.debug(f'payload: {payload}')
+        headers = self._headers
+        headers['Content-Type'] = 'application/json'
+        return self._http_request(method='PUT', url_suffix=f'{URL_SUFFIX}{feature_name}?fields={EDIT_FIELDS}',
+                                  headers=headers, resp_type='json', data=json.dumps(payload))
+
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:67]: removed the dump.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+''' MAIN FUNCTION '''
+
+
+def main() -> None:
+
+    api_key = demisto.params().get('api_key', {}).get('password', {})
+    base_url = urljoin(demisto.params()['url'], '/api/v1')
+    proxy = demisto.params().get('proxy', False)
+    verify = not demisto.params().get('insecure', False)
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:184]: right.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+            message = parse_features(response['features'], DEFAULT_FIELDS.union(
+                fields)) if 'features' in response else parse_feature(response['feature'], DEFAULT_FIELDS.union(fields))
+            human_readable = tableToMarkdown('Aha! get features',
+                                             message,
+                                             removeNull=True)
+    except DemistoException as e:
+        if 'Forbidden' in str(e) or 'Authorization' in str(e):
+            message.append('Authorization Error: make sure API Key is correctly set')
+        else:
+            raise e
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:141]: done.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            fields: Dict fields to update
+        """
+        name = fields.get('name')
+        desc = fields.get('description')
+        status = fields.get('status')
+        payload = {'feature': {'name': name, 'descriptio': desc,
+                   'workflow_status': {'name': status}}}
+        demisto.debug(f'payload: {payload}')
+        headers = self._headers
+        headers['Content-Type'] = 'application/json'
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:64]: done.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    for res in response:
+        curr = parse_feature(res, fields=fields)
+        res_list.extend(curr)
+    return res_list
+
+
+def parse_feature(response: dict, fields: set = DEFAULT_FIELDS) -> List:
+    ret_dict = {}
+    for curr in fields:
+        demisto.info(f'curr: {curr}')
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:82]: done.

File: Packs/AHA/Integrations/AHA/AHA_test.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+from AHA import Client, get_features, edit_feature
+import io
+
+
+apikey = demisto.contentSecrets['AutoFocusTagsFeed']['api_key']
+# test = demisto.params().get('api_key', {}).get('password')
+
+headers = {
+    'Authorization': f'Bearer {apikey}',
+}
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA_test.py:12]: done

File: Packs/AHA/Integrations/AHA/AHA_test.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        mocker.patch.object(client, '_http_request', side_effect=DemistoException(err_msg, res={}))
+
+    return client
+
+
+def util_load_json(path):
+    with io.open(path, mode='r', encoding='utf-8') as f:
+        return json.loads(f.read())
+
+
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA_test.py:34]: done

File: Packs/AHA/Integrations/AHA/AHA.yml [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+  isfetch: false
+  runonce: false
+  script: "-"
+  type: python
+  subtype: python3
+  dockerimage: demisto/python3:3.10.7.33922
+  feed: false
+fromversion: 6.5.0
+tests:
+  - No tests (auto formatted)
```
[the remark this answers is not in the document]
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.yml:140]: done.

File: Packs/AHA/Integrations/AHA/README.md (hunk 1 above)
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/README.md:34]: not sure about this. i have given possible values in command example section.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+        feature = payload.get('feature', {})
+        if field == 'status':
+            workflow_status = {'name': fields[field]}
+            feature['workflow_status'] = workflow_status
+        else:
+            feature[field] = fields[field]
+    return payload
+
+
+def parse_features(response: dict, fields: List) -> List:
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA.py:89]:
```suggestion
def parse_features(features: dict, fields: List = DEFAULT_FIELDS) -> List:
```

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+    res_list = []
+    for res in response:
+        curr = parse_feature(res, fields=fields)
+        res_list.extend(curr)
+    demisto.debug(f'Parsed response fields: {res_list}')
+    return res_list
+
+
+def parse_feature(response: dict, fields: List = DEFAULT_FIELDS) -> List:
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA.py:98]:
```suggestion
def parse_feature(feature: dict, fields: List = DEFAULT_FIELDS) -> List:
```

File: Packs/AHA/Integrations/AHA/AHA.py (hunk 16 above)
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:89]: done.

File: Packs/AHA/Integrations/AHA/AHA.py (hunk 17 above)
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:98]: done.

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                 url: str):
+        super().__init__(base_url=base_url, proxy=proxy, verify=verify, headers=headers)
+        self.url = url
+        self._headers['Content-Type'] = 'application/json'
+
+    def get_features(self,
+                     feature_name: str,
+                     fields: str,
+                     from_date: str,
+                     page: str,
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA.py:38]: `page` isn't int?

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+                     per_page: str) -> Dict:
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA.py:39]: `per_page` isn't in?

File: Packs/AHA/Integrations/AHA/AHA.py [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+    """Tests API connectivity and authentication'"""
+
+    message: str = ''
+    try:
+        result = client.get_features('', '', '2020-01-01', page='1', per_page='1')
+        if result:
+            message = 'ok'
+    except DemistoException as e:
+        if 'Forbidden' in str(e) or 'Authorization' in str(e):
+            message = 'Authorization Error: make sure API Key is correctly set'
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA.py:123]:
```suggestion
            message = 'Authorization Error: make sure that the API Key is setup correctly'
```

File: Packs/AHA/Integrations/AHA/AHA_description.md [hunk 21]
```diff
@@ -0,0 +1,10 @@
+ To access the Aha! service, you need to provide 
+ your Aha! server URL for example: "https://example.com.aha.io/"
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA_description.md:2]:
```suggestion
 your Aha! server URL. For example: "https://example.com.aha.io/"
```

File: Packs/AHA/Integrations/AHA/AHA_description.md [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+ And API key as parameters for the integration.
+ 
+ To generate your personal api_key, visit:
+ https://secure.aha.io/settings/api_keys
+ 
+ The integration supports the following workflows:
+ - Aha! Get features (Lists all features unless a specific feature is requested.)
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA_description.md:9]:
```suggestion
 - Aha! Get features (Lists all features, or a specific feature.)
```

File: Packs/AHA/Integrations/AHA/AHA_description.md [hunk 23]
```diff
@@ -0,0 +1,10 @@
...
+ - Aha! Edit feature (Can Edit a feature name, description and status.)
```
Comment (tkatzir) [inline Packs/AHA/Integrations/AHA/AHA_description.md:10]:
```suggestion
 - Aha! Edit feature (you can Edit a feature name, description, and status.)
```

**Review by tkatzir — APPROVED** [review]
Looks great!
Approved, but please see the minor comments before merging.

File: Packs/AHA/Integrations/AHA/AHA.py (hunk 19 above)
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:39]: it is. but i get it from parameters passed to the command as a string and i pass it to the service API client as a string. so i thought no need in converting it to an integer

File: Packs/AHA/Integrations/AHA/AHA.py (hunk 18 above)
Comment (thefrieddan1) [reply Packs/AHA/Integrations/AHA/AHA.py:38]: same answer as for per_page

**Review by ilappe — APPROVED** [review]

**Comment by ShirleyDenkberg** [comment]
@thefrieddan1 Need to create content for the Pack README.  I sent you information about this in Slack.

**Comment by ShirleyDenkberg** [comment]
@tkatzir @ilappe Doc review completed.

**Pull request merged**